File: CPP/PointGeometry.cpp
// PointGeometry.cpp

#include <cmath> // C/C++ 自带库先引用

#include "ConvertLib.h"
#include "PointGeometry.h"
#include "UUIDGenerator.h"

namespace GeoLib
{
    PointGeometry::PointGeometry(std::shared_ptr<PointModel> point) {
        this->_id = UUIDGenerator::generateUUID();
        this->_type = constHelper::GeometryType::Point;

        if (!point) {
            point = std::make_shared<PointModel>();
        }
        _position = *point;
    }

    PointGeometry::PointGeometry(PointModel point) {
        this->_id = UUIDGenerator::generateUUID();
        this->_type = constHelper::GeometryType::Point;
        _position = point;
    }

    PointGeometry::PointGeometry() {
        this->_id = UUIDGenerator::generateUUID();
        this->_type = constHelper::GeometryType::Point;
    }

    std::string PointGeometry::getType(constHelper::GeometryType _type) { // 添加方法参数
        return ConvertLib::geometryTypeToString(_type);
    }

    std::string PointGeometry::getId() const {
        return _id;
    }

    void PointGeometry::setId(const std::string& id) {
        _id = id;
    }

    PointModel PointGeometry::getPosition() {
        return _position;
    }

    void PointGeometry::setPosition(const PointModel& position) {
        _position = position;
    }

    float PointGeometry::toPointDistance(const PointModel& point) const {
        float distance = 0;
        distance = std::sqrt(std::pow((point.getX() - _position.getX()), 2) + std::pow((point.getY() - _position.getY()), 2) + std::pow((point.getZ() - _position.getZ()), 2));
        return distance;
    }
}

File: CSharp/PointGeometry.cs
//=====================================================================
// 所述类库：基本几何体算法类库
// 版权声明：2019 九州创智科技有限公司  All Rights Reserved.
// 更新履历：2019.6.14 崔艳龙 创建
//============================================


using System;

namespace GeoLib
{
    public class PointGeometry
    {
        private string _type;
        private string _id;

        private PointModel _position;

        /// <summary>
        /// 点构造函数
        /// </summary>
        /// <param name="x">X坐标</param>
        /// <param name="y">Y坐标</param>
        /// <param name="z">Z坐标</param>
        public PointGeometry(PointModel point)
        {
            this._id = Guid.NewGuid().ToString();
            this._type = constHelper.GeometryType.Point.ToString();

            if (point == null) point = new PointModel();
            _position = point;
        }

        //public PointGeometry(string jsonStr = "")
        //{
        //    this._id = Guid.NewGuid().ToString();
        //    if (!string.IsNullOrEmpty(jsonStr)) FromJsonString(jsonStr);
        //}

        public string type
        {
            get
            {
                return _type;
            }            
        }

        public string Id
        {
            get
            {
                return _id;
            }
            set
            {
                _id = value;
            } 
        }

        public PointModel position
        {
            get
            {
                return _position;
            }

            set
            {
                _position = value;
            }
        }

        /// <summary>
        /// 同另一个点距离
        /// </summary>
        /// <param name="point">另一个点坐标</param>
        /// <returns></returns>
        public float ToPointDistance(PointModel point)
        {
            float distance = 0;
            distance = (float)Math.Sqrt((Math.Pow((point.x - _position.x), 2) + Math.Pow((point.y - _position.y), 2) + Math.Pow((point.x - _position.z), 2)));

            return distance;
        }
    }
}
